requirements.txt

```diff
@@ -1,3 +1,5 @@
 aws-cdk-lib==2.139.0
+aws-cdk.aws-scheduler-alpha==2.139.0a0
```

Comment:
better to do a range than to pin to a specific version,  i would go with `~=2.139` for all aws-cdk* dependencies.  so like this.. `aws-cdk.aws-scheduler-alpha~=2.139`

Reply:
Could you please describe why using ranges would be better?

Reply:
Ranges are typically better because then you may do a re-install of the requirements to pull in new patch versions for a particular library. Pinning to a specific version means you need to be intentional when you want to upgrade (Not a bad thing), but it's another thing to think about.

Depending on the library I'd recommend either:

1. Pin to a major version and let minor/patches be upgraded
2. Pin to a specific minor version and let patches be upgraded -> The suggestion here
3. Enforce a an upper bound of the range, such as preventing the next major version from being used

Reply:
I think the suggestion here covers all 3 items?

Reply:
`~=` is the Compatible release:
https://peps.python.org/pep-0440/#compatible-release

Which as I am reading looks like it would be equivalent to:

`>= 2.139, == 2.*`

Which, as you mention actually does cover all 3 I mentioned above.

Reply:
Note that packages are not automatically updated when installing them from a lock file.

This project doesn't have a lock file and we should add one.

In application projects, I always use a lock file to pin dependency versions for reproducibility and stability. Dependencies are then updated periodically to reduce the risk of unexpected breakage.

I’ve had this discussion a few times, so I created a [Confluence page](https://sagebionetworks.jira.com/wiki/spaces/MONOREPO/pages/3794763782/Managing+Dependency+Versions)—with help from ChatGPT—to outline the issue and provide guidance, especially for those working in the monorepo.

If you have time, I’d appreciate your feedback before sharing it broadly.

Reply:
would be a good discussion in slack but in general you can do a hybrid of ranges and pinning with tools like pipenv or poetry.  pipenv is the more lightweight tool and maybe all you need.